### DISCUSSION TITLE
Abaqus UMAT in MOOSE

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#21201] [Q&A Modules: Solid mechanics] [answered]
Hello,

I was wondering how one can compile a multi-file UMAT with MOOSE. If I'm not mistaken, I need a single .plugin file to use with AbaqusUMATStress.C

Another question I have is what exactly is a .plugin file? When I manually compile my UMAT I get an umat.o and umat.so. Is it possible to manually compile the UMAT and use one of these files with MOOSE?

**Comment by GiudGiud** [comment]
@dschwen

**Comment by ngrilli** [comment, marked as answer]
Dear @Flolaffel and @GiudGiud 

It depends on the specific nature of your UMAT.

An example of what I do is reported here:
https://github.com/ngrilli/c_pfor_am/tree/main/test/plugins/Bristol_CP
and see the corresponding Makefile commands here:
https://github.com/ngrilli/c_pfor_am/blob/main/Makefile

The UMAT is divided into modules
and the main file is the one that contains UEXTERNALDB and UMAT subroutines.
Modules are imported in the various subroutines using the "use" keyword.

"include" keyword is also necessary, see here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/plugins/Bristol_CP/BRISTOL.f

I need first to compile the modules one by one using:
gfortran -c -free name_of_module.f
in the order of dependency, meaning a module A dependent on B will be compiled after B.
This will generate .mod files.

Once all the modules are compiled, run "make" as usual in the MOOSE App.

The main file, which is the one with the UMAT inside is not a module and will be compiled by make.
If you don't compile all the modules, when you "make", you get an error saying missing module.

I believe if you don't use modules, it is sufficient to "include" the necessary files,
like here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/plugins/Oxford_CP/umat.f
but have not tested this extensively.
In this case, just go for "make" directly.

Both procedures, with/without modules, will generate one .plugin file for each fortran file.
In the MOOSE input file, you will always point to the .plugin file corresponding to the fortran file
that contains the UMAT or UEXTERNALDB respectively.

An example input file for the previous code is here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/umat/umat_visualise_SDV.i

Best Regards,
Nicolò

**Reply by Flolaffel** [reply]
Thanks a lot for your answer. While trying it out, I got an error that says 

`Can't open included file 'xxx.f90'`

Does that mean that MOOSE cannot handle free format fortran?

**Reply by Flolaffel** [reply]
@ngrilli

**Comment by ngrilli** [comment]
@Flolaffel 
about your second question, I am not sure.
If you put the fortran file inside the ./test/plugins/ folder
of a MOOSE App, the "make" command will compile them and generate the .plugin files.
So, you don't need to compile the UMAT externally.
Nicolò

**Reply by Flolaffel** [reply]
Well since I already have an externally compiled version I was just wondering whether there is a way to use it with MOOSE.

And if my assumption that MOOSE cannot handle free format is correct, it would safe me the work to convert it to fixed.

**Reply by Flolaffel** [reply]
@ngrilli

**Reply by GiudGiud** [reply]
I think it does. You may just have to pass additional compile flags
In this thread https://github.com/idaholab/moose/discussions/21476
Jesse mentions the PLUGIN_FLAGS in the Makefile
